File: src/patterns.cpp
//
// Created by stijn on 6/8/23.
//

#include <vector>
#include <algorithm>

#include "patterns.h"
#include "motif.h"
#include "typing.h"

PatternMiner::PatternMiner(double minsup, int min_len, int max_len, double max_overlap) : minsup(minsup), min_len(min_len), max_len(max_len),
                                                                              max_overlap(max_overlap), min_freq(0.0),
                                                                              k(2), k_motifs(), k_1_motifs(){}

void PatternMiner::mine(const DiscreteDB& sequences)
{
    // Frequency is easier to check than support
    min_freq = static_cast<double>(sequences.size()) * minsup;

    // Mine 1-patterns separately from longer patterns
    mine_1_patterns(sequences);

    // If there were no frequent k-patterns, there can be no frequent (k+1)-patterns; stop
    for (k = 2; (!k_1_motifs.empty() and ((not max_len or k <= max_len))); k++) {
        // Generate candidate k-patterns from frequent (k-1)-patterns, find their occurrences and remove infrequent candidates
        for (auto& candidate : get_candidates()) {
            Motif motif { candidate };
            find_occurrences(candidate, sequences, motif);

            // Add frequent patterns to frequent vector
            if (static_cast<double>(motif.get_indexes().size()) >= min_freq) {
                k_motifs.insert(std::pair(candidate, motif));
                frequent.push_back(candidate);
            }
        }
        k_1_motifs = k_motifs;
        k_motifs.clear();
    }

    // Remove patterns that are too short or overlap too much with a longer pattern
    remove_redundant();
}

void PatternMiner::mine_1_patterns(const DiscreteDB& sequences)
{
    // Scan database once to record indexes of 1-patterns
    for (size_t i = 0; i < sequences.size(); i++) {
        for (size_t j = 0; j < sequences[i].size(); j++) {
            Pattern item = {sequences[i][j]};
            if (k_1_motifs.count(item) == 0) {
                k_1_motifs.insert(std::pair(item, Motif(item)));
            }
            k_1_motifs.at(item).record_index(i, j);
        }
    }

    // Add frequent patterns to frequent vector
    for (auto& [pattern, motif] : k_1_motifs) {
        if (static_cast<double>(motif.get_indexes().size()) >= min_freq) {
            frequent.push_back(pattern);
        }
    }
}

std::vector<Pattern> PatternMiner::get_candidates()
{
    std::vector<Pattern> candidates { };
    for (auto& [p1, m1] : k_1_motifs) {
        for (auto& [p2, m2] : k_1_motifs) {
            // Check if p1[1:] == p2[:-1]
            if (std::equal(p1.begin()+1, p1.end(), p2.begin(), p2.end()-1)) {
                // Construct candidate by combing p1 and p2
                Pattern candidate { p1 };
                candidate.push_back(p2.back());
                candidates.push_back(candidate);
            }
        }
    }
    return candidates;
}

void PatternMiner::find_occurrences(const Pattern& candidate, const DiscreteDB& sequences, Motif& motif)
{
    // Find candidate via its first parent
    Pattern parent(candidate.begin(), candidate.end() - 1);
    for (auto& [seq, indexes] : k_1_motifs.at(parent).get_indexes()) {
        for (auto index : indexes) {
            // If index + length is larger than sequence size, candidate can never be present at index
            // Omitting this check leads to a bug if start of next sequence would complete the pattern
            if (index + k > sequences[seq].size()) {
                continue;
            }

            Pattern possible_match(sequences[seq].begin() + index, sequences[seq].begin() + index + k);
            if (possible_match == candidate) {
                motif.record_index(seq, index);
            }
        }
    }
}

void PatternMiner::remove_redundant()
{
    // Remove patterns that are too short
    frequent.erase(
            std::remove_if(
                    frequent.begin(),
                    frequent.end(),
                    [this](const std::vector<char>& v) {
                        return v.size() < min_len;
                    }
            ),
            frequent.end()
    );

    // If max_overlap == 1, no overlap is too high
    if (max_overlap == 1) {
        return;
    }

    // Create a copy of frequent to edit frequent in the loops
    std::vector<Pattern> patterns = frequent;

    // Sort the copy in decreasing order of vector lengths
    std::sort(patterns.begin(), patterns.end(),
              [](const std::vector<char>& a, const std::vector<char>& b) {
                  return a.size() > b.size();
              });

    // Remove patterns for which more than max_overlap% is overlapping from frequent
    std::vector<Pattern> removed;
    for (const auto& p1 : patterns) {
        // Check if p1 was not already removed
        if (is_p_in_vec(p1, removed)) {
            continue;
        }
        for (const auto& p2 : patterns) {
            if (p2.size() >= p1.size() or is_p_in_vec(p2, removed)) {
                continue;
            }

            // Check if shorter patterns overlaps too much with larger pattern
            if (lcs(p1, p2) / static_cast<double>(p2.size()) > max_overlap) {
                auto pos = std::find(frequent.begin(), frequent.end(), p2);
                frequent.erase(pos);
                removed.push_back(p2);
            }
        }
    }
}

bool is_p_in_vec(const Pattern& p, const std::vector<Pattern>& vec)
{
    return std::find(vec.begin(), vec.end(), p) != vec.end();
}

double lcs(const Pattern& p1, const Pattern& p2) {
    size_t n = p1.size();
    size_t m = p2.size();

    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(m + 1, 0));

    for (size_t i = 1; i <= n; i++) {
        for (size_t j = 1; j <= m; j++) {
            if (p1[i - 1] == p2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    return static_cast<double>(dp[n][m]);
}

File: frm/_frm_py/patterns.py
"""Patterns module.

This module defines the PatternMiner class, a class that takes a collection of
sequences and mines frequent and maximal patterns from it. The patterns
can be filtered to have a certain minimum length.
"""
from .motif import Motif


class PatternMiner:
    """Mine patterns from a collection of sequences with PatternMiner.

    Parameters
    ----------
    minsup : float
        The minimum support for a pattern.
    
    Attributes
    ----------
    frequent : dict
        Dictionary of frequent motifs with string patterns as keys and Motif objects as values.
    """
    def __init__(
            self,
            minsup: float = 0.5,
            min_len: int = 3,
            max_len: int = 0,
            max_overlap: float = 0.9):
        self.minsup = minsup
        self.min_len = min_len
        self.max_len = max_len
        self.max_overlap = max_overlap

        self.frequent = {}

        # Frequency is easier to check than support
        self._min_freq = 0

        # Keep track of length we're currently mining
        self._k = 2

        # self.patterns[k] contains a list of patterns of length k
        self._patterns = [[], []]

    def mine(self, sequences):
        """Mine sequence motifs.

        Starting with patterns of length 1 (1-patterns), candidates are
        generated and checked for sufficient support. Then, k-patterns
        are generated based on (k-1)-patterns.

        sequences : list
            Collection of sequences with discrete values.
        """
        self._min_freq = len(sequences) * self.minsup

        # Mine 1-patterns separately from longer patterns
        self.mine_1_patterns(sequences)

        # If there were no frequent k-patterns, there can be no 
        # frequent (k+1)-patterns; stop
        while self._patterns[self._k - 1] and (not self.max_len or self._k <= self.max_len):
            self._patterns.append([])

            for candidate in self.get_candidates():
                # Count candidate occurrences
                self.frequent[candidate] = Motif(candidate)

                # Find candidate in sequences parent occurs in
                for seq, indexes in self.frequent[candidate[:-1]].indexes.items():
                    sequence = sequences[seq]
                    for index in indexes:
                        if sequence[index : index+self._k] == candidate:
                            self.frequent[candidate].record_index(seq, index)
                
                # Check if candidate complies with minimum support
                self.prune_infrequent(candidate)

            self._k += 1

        self.remove_redundant()
    
    def remove_redundant(self):
        """Remove redundant patterns.
        
        Not all frequent patterns that comply with the minimum support
        are interesting. For example, 1-patterns are too trivial to be
        considered interesting.
        Furthermore, mining only maximal patterns can leave patterns
        that are trivial because they are almost contained in other
        patterns. For instance, if aaaaaa is a frequent pattern, aabaa
        could be considered trivial, as almost all of its contents are
        contained in the longer pattern.

        Parameters
        ----------
        min_len : int, optional.
            Minimum length of a pattern to be considered interesting.
        o : float, optional.
            Maximal redundancy ratio, calculated by
            length(longest common subsequence) / length(shortest pattern).
            If the redundancy ratio exceeds `max_overlap`, the shortest pattern is
            pruned.
        """
        # Prune from frequent patterns
        self.frequent = {k: v for k, v in self.frequent.items() if len(k) >= self.min_len}

        # If max_overlap == 1, no overlap is too high
        if self.max_overlap == 1:
            return

        # Prune patterns for which x% is overlapping from frequent
        patterns = list(self.frequent)
        patterns.sort(key=lambda x: (len(x), x), reverse=True)
        pruned = []
        
        for p1 in patterns:
            if p1 in pruned:
                continue
            for p2 in patterns:
                n, m = len(p1), len(p2)
                # Only check unseen patterns and pairs
                if m > n or p1 == p2 or p2 in pruned:
                    continue

                # Check if shorter pattern consists mostly of lcs
                if self.lcs(p1, p2, n, m) / m > self.max_overlap:
                    self.frequent.pop(p2, 0)
                    pruned.append(p2)

    def lcs(self, p1: str, p2: str, n: int, m: int) -> int:
        """Longest common subsequence.
        
        Find the length of the longest sequence that is contained in two
        patterns. Uses dynamic programming algorithm adapted from
        https://www.geeksforgeeks.org/longest-common-subsequence-dp-4/

        Parameters
        ----------
        p1 : str
            String representation of first pattern.
        p2 : str
            String representation of second pattern.
        n : int
            Length of first pattern.
        m : int
            Length of second pattern.
        
        Returns
        -------
        res : int
            Length of the longest common subsequence of `p1` and `p2`.
        """
        # declaring the array for storing the dp values
        L = [[0]*(m+1) for i in range(n+1)]

        """Following steps build L[n+1][m+1] in bottom up fashion
        Note: L[i][j] contains length of LCS of X[0..i-1]
        and Y[0..j-1]"""
        for i in range(n+1):
            for j in range(m+1):
                if i == 0 or j == 0:
                    L[i][j] = 0
                elif p1[i-1] == p2[j-1]:
                    L[i][j] = L[i-1][j-1]+1
                else:
                    L[i][j] = max(L[i-1][j], L[i][j-1])

        # L[n][m]] contains the length of LCS of X[0..n-1] & Y[0..m-1]
        return L[n][m]
    
    def get_candidates(self) -> list[str]:
        """Use frequent (k-1)-patterns to generate candidate k-patterns.
        
        Candidates are generated by joining two patterns from L[k-1]
        if pattern_1[1:] == pattern_2[:-1].

        Returns
        -------
        candidates : list[str]
            Set of candidate patterns that have a length of k.
        """
        patterns = self._patterns[self._k - 1]
        return [p1 + p2[-1] for p1 in patterns for p2 in patterns if p1[1:] == p2[:-1]]
    
    def mine_1_patterns(self, sequences):
        """Mine frequent 1-patterns.
        
        One scan is made over the sequences to map 1-pattern indexes.

        sequences : list
            Collection of sequences with discrete values.
        """
        # Divide sequences into indexes
        for i, sequence in enumerate(sequences):
            for j, item in enumerate(sequence):
                if item not in self.frequent:
                    self.frequent[item] = Motif(item)
                self.frequent[item].record_index(i, j)

        # Prune infrequent patterns
        for a in list(self.frequent.keys()):
            self.prune_infrequent(a)
    
    def prune_infrequent(self, pattern: str):
        """Prune infrequent patterns.
        
        - Prunes patterns with a too low support;
        - Adds frequent patterns to list of frequent patterns

        Parameters
        ----------
        pattern : str
            The pattern that should be processed.
        """
        if len(self.frequent[pattern].indexes) < self._min_freq:
            # Delete indexes of this infrequent pattern
            self.frequent.pop(pattern)
        else:
            # Add to list of frequent k-patterns
            self._patterns[len(pattern)].append(pattern)
